### PR TITLE
Add support for torch.is_nonzero() and tests for torch.is_nonzero()

### PR DESCRIPTION
**Pull request by AnishPahilajani**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

- Adds support for torch.is_nonzero() on Spyre backend
- Adds unit tests

#### Which issue(s) this PR is related to:

Depends on https://github.com/torch-spyre/torch-spyre/pull/786 (hard dependency)

### CODE REVIEW

**Review by ani300 — CHANGES_REQUESTED** [review]
the decomposition needs some work, the tests look almost good

File: tests/inductor/test_inductor_ops.py [hunk 1]
```diff
@@ -1422,6 +1422,25 @@ class TestOps(unittest.TestCase, metaclass=ParameterizedTestMeta):
                 ),
             },
         },
+        ("test_is_nonzero", "test_is_nonzero_cpu"): {
+            "param_sets": {
+                "float16_true": (torch.tensor([3.14], dtype=torch.float16),),
+                "float16_false": (torch.tensor([0.0], dtype=torch.float16),),
+                # "float32_true": (torch.tensor([2.71828], dtype=torch.float32),), # # TODO: FP32 will be supported soon
+                # "float32_true": (torch.tensor([0.0], dtype=torch.float32),), # # TODO: FP32 will be supported soon
+                # "negative_true": (torch.tensor([-1.0], dtype=torch.float32),), # TODO: FP32 will be supported soon
```
[the remark this answers is not in the document]
Comment (avery-blanchard) [reply tests/inductor/test_inductor_ops.py]: This op (and the other comparison operations) should have fp32 support. We should add this to the list here!
If you hit any issues, let me know and we can check on it.

[the remark this answers is not in the document]
Comment (AnishPahilajani) [reply tests/inductor/test_inductor_ops.py]: createde an issue for this: https://github.com/torch-spyre/torch-spyre/issues/1543

**Comment by mudhakar** [comment]
@marnold-ibm can you review this please ^^

File: tests/inductor/test_inductor_ops.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (avery-blanchard) [reply tests/inductor/test_inductor_ops.py]: @AnishPahilajani - there is support for this in the backend already. If you hit issues here, then it is a bug. Are you seeing errors running this operation in fp32?

**Comment by marnold-ibm** [comment]
@mudhakar sure happy to, but @ani300 and @avery-blanchard have discussions in flight so would like to see the resolution of those before weighing in?

File: tests/inductor/test_inductor_ops.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AnishPahilajani) [reply tests/inductor/test_inductor_ops.py]: yes I am seeing errors when running with fp32.

```
=========================== short test summary info ============================
FAILED test_inductor_ops__oot_wrapper.py::TestOpsPRIVATEUSE1::test_is_nonzero_float32_false_spyre
FAILED test_inductor_ops__oot_wrapper.py::TestOpsPRIVATEUSE1::test_is_nonzero_float32_true_spyre
FAILED test_inductor_ops__oot_wrapper.py::TestOpsPRIVATEUSE1::test_is_nonzero_negative_true_spyre
===== 3 failed, 839 passed, 127 xfailed, 78 warnings in 518.49s (0:08:38) ======
```

[the remark this answers is not in the document]
Comment (avery-blanchard) [reply tests/inductor/test_inductor_ops.py]: @AnishPahilajani are they accuracy or compilation errors?

[the remark this answers is not in the document]
Comment (AnishPahilajani) [reply tests/inductor/test_inductor_ops.py]: They look like compile errors. you can see the failures in the github PR tests https://github.com/torch-spyre/torch-spyre/actions/runs/26064830357/job/76633198490?pr=893

[the remark this answers is not in the document]
Comment (avery-blanchard) [reply tests/inductor/test_inductor_ops.py]: Got it. Lets mark these as xfail for now. Followed up on slack.

**Comment by isuruf** [comment]
Are the changes to decompositions.py needed now? It seems to work on main.

**Comment by dgrove-oss** [comment]
Please resolve the conflicts and we will attempt to merge before it goes out of date again (ping me when it is ready).

I asked the code review agent to analyze the test coverage provided.  The response is below.  Consider this when resolving the conflicts.

-----

Are the tests providing new coverage or are they redundant?

Mostly new coverage, with one caveat.

New coverage the PR adds:

1. test_is_nonzero_cpu (parameterized) — This is the core compiled-path test. It's genuinely new: there are no existing compiled-path tests for is_nonzero anywhere in the test suite. The param sets cover float16, bfloat16, bool, int64, and a from_computation case (multiply then check nonzero). This exercises the Spyre backend implementation from PR #786.
2. test_is_nonzero_error_cases — Tests that multi-element tensors raise RuntimeError in both eager and compiled contexts. The compiled-path multi-element error check is new.

Partially redundant:

- TestTorch::test_is_nonzero already exists in tests/configs/upstream_tests_beta/test_torch.yaml:80. That upstream test from PyTorch's own test_torch.py runs on the Spyre device and covers basic nonzero behavior. The new PR's eager-mode error cases (empty tensor, multi-element tensor) likely overlap with what TestTorch::test_is_nonzero already exercises.

The empty-tensor error case is dead code:

The entire compiled Spyre path for the empty-tensor case (test_is_nonzero_error_cases lines ~4505–4518) is commented out with a # TODO. So the only coverage for the empty tensor is the raw PyTorch eager call — which tests PyTorch itself, not the Spyre backend. That block contributes no Spyre-specific coverage.

Summary: The parameterized test_is_nonzero_cpu tests are the valuable new addition — compiled-path coverage that doesn't exist elsewhere. The error-case test is mostly redundant with the upstream suite and the commented-out Spyre compiled path for empty tensors should either be implemented or the test should call that out more clearly.

File: tests/inductor/test_inductor_ops.py [hunk 2]
```diff
@@ -2465,6 +2465,25 @@ def setUp(self):
                 ),
             },
         },
+        ("test_is_nonzero", "test_is_nonzero_cpu"): {
+            "param_sets": {
+                "float16_true": (torch.tensor([3.14], dtype=torch.float16),),
+                "float16_false": (torch.tensor([0.0], dtype=torch.float16),),
+                # "float32_true": (torch.tensor([2.71828], dtype=torch.float32),), # TODO
+                # "float32_false": (torch.tensor([0.0], dtype=torch.float32),), # TODO
+                # "negative_true": (torch.tensor([-1.0], dtype=torch.float32),), # TODO
```
Comment (ani300) [inline tests/inductor/test_inductor_ops.py]: can we mark these as xfail through the expect_fail API or do we need to skip them due to crashes?

Comment (AnishPahilajani) [reply tests/inductor/test_inductor_ops.py]: marked as xfail

**Review by ani300 — APPROVED** [review]
